In [222]:
import numpy as np
import pandas as pd
import time

In [205]:
data = pd.read_csv('botswana.tsv', sep = '\t', header = 0)

In [206]:
data.head()

ceb  age  educ    religion  idlnchld  knowmeth  usemeth  evermarr  agefm  \
0    0   18    10    catholic       4.0       1.0      1.0         0    NaN   
1    2   43    11  protestant       2.0       1.0      1.0         1   20.0   
2    0   49     4      spirit       4.0       1.0      0.0         1   22.0   
3    0   24    12       other       2.0       1.0      0.0         0    NaN   
4    3   32    13       other       3.0       1.0      1.0         1   24.0   

   heduc  urban  electric  radio   tv  bicycle  
0    NaN      1       1.0    1.0  1.0      1.0  
1   14.0      1       1.0    1.0  1.0      1.0  
2    1.0      1       1.0    1.0  0.0      0.0  
3    NaN      1       1.0    1.0  1.0      1.0  
4   12.0      1       1.0    1.0  1.0      1.0

Давайте проанализируем данные опроса 4361 женщин из Ботсваны:

О каждой из них мы знаем:

    сколько детей она родила (признак ceb)
    возраст (age)
    длительность получения образования (educ)
    религиозная принадлежность (religion)
    идеальное, по её мнению, количество детей в семье (idlnchld)
    была ли она когда-нибудь замужем (evermarr)
    возраст первого замужества (agefm)
    длительность получения образования мужем (heduc)
    знает ли она о методах контрацепции (knowmeth)
    использует ли она методы контрацепции (usemeth)
    живёт ли она в городе (urban)
    есть ли у неё электричество, радио, телевизор и велосипед (electric, radio, tv, bicycle)

Давайте научимся оценивать количество детей ceb по остальным признакам.

Загрузите данные и внимательно изучите их. Сколько разных значений принимает признак religion?

In [207]:
pd.unique(data['religion'])

array(['catholic', 'protestant', 'spirit', 'other'], dtype=object)

Question 1: 4

Во многих признаках есть пропущенные значения. Сколько объектов из 4361 останется, если выбросить все, содержащие пропуски? 

In [208]:
data_without_NaN = data.dropna()

In [209]:
data.shape

(4361, 15)

In [210]:
data_without_NaN.shape

(1834, 15)

Question 2: 1834

Обработка пропущенных значений в heduc.
Сколько осталось пропущенных значений в признаке heduc?

In [211]:
data.loc[:,'nevermarr'] = pd.Series(np.zeros(data.shape[0]), index=data.index)

In [212]:
data.shape

(4361, 16)

In [213]:
data.head()

ceb  age  educ    religion  idlnchld  knowmeth  usemeth  evermarr  agefm  \
0    0   18    10    catholic       4.0       1.0      1.0         0    NaN   
1    2   43    11  protestant       2.0       1.0      1.0         1   20.0   
2    0   49     4      spirit       4.0       1.0      0.0         1   22.0   
3    0   24    12       other       2.0       1.0      0.0         0    NaN   
4    3   32    13       other       3.0       1.0      1.0         1   24.0   

   heduc  urban  electric  radio   tv  bicycle  nevermarr  
0    NaN      1       1.0    1.0  1.0      1.0        0.0  
1   14.0      1       1.0    1.0  1.0      1.0        0.0  
2    1.0      1       1.0    1.0  0.0      0.0        0.0  
3    NaN      1       1.0    1.0  1.0      1.0        0.0  
4   12.0      1       1.0    1.0  1.0      1.0        0.0

In [214]:
for i in range(4361):
    if data['agefm'].isna()[i] == True:
        data['nevermarr'][i] = 1

C:\Users\oldst\Anaconda2\lib\site-packages\ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  This is separate from the ipykernel package so we can avoid doing imports until


In [215]:
data.head()

ceb  age  educ    religion  idlnchld  knowmeth  usemeth  evermarr  agefm  \
0    0   18    10    catholic       4.0       1.0      1.0         0    NaN   
1    2   43    11  protestant       2.0       1.0      1.0         1   20.0   
2    0   49     4      spirit       4.0       1.0      0.0         1   22.0   
3    0   24    12       other       2.0       1.0      0.0         0    NaN   
4    3   32    13       other       3.0       1.0      1.0         1   24.0   

   heduc  urban  electric  radio   tv  bicycle  nevermarr  
0    NaN      1       1.0    1.0  1.0      1.0        1.0  
1   14.0      1       1.0    1.0  1.0      1.0        0.0  
2    1.0      1       1.0    1.0  0.0      0.0        0.0  
3    NaN      1       1.0    1.0  1.0      1.0        1.0  
4   12.0      1       1.0    1.0  1.0      1.0        0.0

In [216]:
del data['evermarr']
data.head()

ceb  age  educ    religion  idlnchld  knowmeth  usemeth  agefm  heduc  \
0    0   18    10    catholic       4.0       1.0      1.0    NaN    NaN   
1    2   43    11  protestant       2.0       1.0      1.0   20.0   14.0   
2    0   49     4      spirit       4.0       1.0      0.0   22.0    1.0   
3    0   24    12       other       2.0       1.0      0.0    NaN    NaN   
4    3   32    13       other       3.0       1.0      1.0   24.0   12.0   

   urban  electric  radio   tv  bicycle  nevermarr  
0      1       1.0    1.0  1.0      1.0        1.0  
1      1       1.0    1.0  1.0      1.0        0.0  
2      1       1.0    1.0  0.0      0.0        0.0  
3      1       1.0    1.0  1.0      1.0        1.0  
4      1       1.0    1.0  1.0      1.0        0.0

In [217]:
%time
for i in range(4361):
    if data['agefm'].isna()[i] == True:
        data['agefm'][i] = 0
%time

Wall time: 0 ns


C:\Users\oldst\Anaconda2\lib\site-packages\ipykernel_launcher.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  after removing the cwd from sys.path.


Wall time: 0 ns


In [218]:
data.head()

ceb  age  educ    religion  idlnchld  knowmeth  usemeth  agefm  heduc  \
0    0   18    10    catholic       4.0       1.0      1.0    0.0    NaN   
1    2   43    11  protestant       2.0       1.0      1.0   20.0   14.0   
2    0   49     4      spirit       4.0       1.0      0.0   22.0    1.0   
3    0   24    12       other       2.0       1.0      0.0    0.0    NaN   
4    3   32    13       other       3.0       1.0      1.0   24.0   12.0   

   urban  electric  radio   tv  bicycle  nevermarr  
0      1       1.0    1.0  1.0      1.0        1.0  
1      1       1.0    1.0  1.0      1.0        0.0  
2      1       1.0    1.0  0.0      0.0        0.0  
3      1       1.0    1.0  1.0      1.0        1.0  
4      1       1.0    1.0  1.0      1.0        0.0

In [221]:
for i in range(4361):
    if data['nevermarr'][i] == 1:
        data['heduc'][i] = -1

C:\Users\oldst\Anaconda2\lib\site-packages\ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  This is separate from the ipykernel package so we can avoid doing imports until


In [223]:
data.describe()['ceb'][0]  - data.describe()['heduc'][0]

123.0

Question 3: 123

Избавимся от оставшихся пропусков.

Для признаков idlnchld, heduc и usemeth проведите операцию, аналогичную предыдущей: создайте индикаторы пропусков по этим признакам (idlnchld_noans, heduc_noans, usemeth_noans), замените пропуски на нехарактерные значения (cidlnchld=−1, cheduc2=−2(значение -1 мы уже использовали), cusemeth=−1.

Остались только пропуски в признаках knowmeth, electric, radio, tv и bicycle. Их очень мало, так что удалите объекты, на которых их значения пропущены.

Какого размера теперь наша матрица данных? Умножьте количество строк на количество всех столбцов (включая отклик ceb).

In [224]:
data.loc[:,'idlnchld_noans'] = pd.Series(np.zeros(data.shape[0]), index=data.index)
data.loc[:,'heduc_noans'] = pd.Series(np.zeros(data.shape[0]), index=data.index)
data.loc[:,'usemeth_noans'] = pd.Series(np.zeros(data.shape[0]), index=data.index)
start_time = time.clock()
for i in range(4361):
    if data['idlnchld'].isna()[i] == True:
        data['idlnchld_noans'][i] = 1
        data['idlnchld'][i] = -1
    if data['heduc'].isna()[i] == True:
        data['heduc_noans'][i] = 1
        data['heduc'][i] = -2
    if data['usemeth'].isna()[i] == True:
        data['usemeth_noans'][i] = 1
        data['usemeth'][i] = -1
print time.clock() - start_time, "seconds"

C:\Users\oldst\Anaconda2\lib\site-packages\ipykernel_launcher.py:10: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  # Remove the CWD from sys.path while we load stuff.
C:\Users\oldst\Anaconda2\lib\site-packages\ipykernel_launcher.py:11: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  # This is added back by InteractiveShellApp.init_path()
C:\Users\oldst\Anaconda2\lib\site-packages\ipykernel_launcher.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  import sys
C:\Users\oldst\Anaconda2\li

22.4196202822 seconds


In [225]:
data.head()

ceb  age  educ    religion  idlnchld  knowmeth  usemeth  agefm  heduc  \
0    0   18    10    catholic       4.0       1.0      1.0    0.0   -1.0   
1    2   43    11  protestant       2.0       1.0      1.0   20.0   14.0   
2    0   49     4      spirit       4.0       1.0      0.0   22.0    1.0   
3    0   24    12       other       2.0       1.0      0.0    0.0   -1.0   
4    3   32    13       other       3.0       1.0      1.0   24.0   12.0   

   urban  electric  radio   tv  bicycle  nevermarr  idlnchld_noans  \
0      1       1.0    1.0  1.0      1.0        1.0             0.0   
1      1       1.0    1.0  1.0      1.0        0.0             0.0   
2      1       1.0    1.0  0.0      0.0        0.0             0.0   
3      1       1.0    1.0  1.0      1.0        1.0             0.0   
4      1       1.0    1.0  1.0      1.0        0.0             0.0   

   heduc_noans  usemeth_noans  
0          0.0            0.0  
1          0.0            0.0  
2          0.0            0.0  
3          0.0            0.0  
4          0.0            0.0

In [226]:
data.shape

(4361, 18)

In [227]:
data = data.dropna()

In [228]:
data.shape

(4348, 18)

In [229]:
print data.shape[0]*data.shape[1]

78264


Question 4: 78264

Постройте регрессию количества детей ceb на все имеющиеся признаки методом smf.ols, как в разобранном до этого примере. Какой получился коэффициент детерминации R^2? Округлите до трёх знаков после десятичной точки.

In [242]:
import statsmodels.formula.api as smf

In [243]:
m1 = smf.ols('ceb ~ age + educ + religion + idlnchld + knowmeth + usemeth + agefm + heduc + urban + electric +'\
                    'radio + tv + bicycle + nevermarr + idlnchld_noans + heduc_noans + usemeth_noans', 
             data=data)
fitted = m1.fit()
print fitted.summary()

                            OLS Regression Results                            
Dep. Variable:                    ceb   R-squared:                       0.644
Model:                            OLS   Adj. R-squared:                  0.643
Method:                 Least Squares   F-statistic:                     412.5
Date:                Wed, 03 Jul 2019   Prob (F-statistic):               0.00
Time:                        11:19:27   Log-Likelihood:                -7732.1
No. Observations:                4348   AIC:                         1.550e+04
Df Residuals:                    4328   BIC:                         1.563e+04
Df Model:                          19                                         
Covariance Type:            nonrobust                                         
                             coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------------
Intercept                 -1

Question 5: 0.644

Question 6: 3

In [244]:
sms.het_breuschpagan(fitted.resid, fitted.model.exog)[1]

1.1452927633437192e-225

In [245]:
m2 = smf.ols('ceb ~ age + educ + religion + idlnchld + knowmeth + usemeth + agefm + heduc + urban + electric +'\
                    'radio + tv + bicycle + nevermarr + idlnchld_noans + heduc_noans + usemeth_noans', 
             data=data)
fitted = m2.fit(cov_type='HC1')
print fitted.summary()

                            OLS Regression Results                            
Dep. Variable:                    ceb   R-squared:                       0.644
Model:                            OLS   Adj. R-squared:                  0.643
Method:                 Least Squares   F-statistic:                     345.0
Date:                Wed, 03 Jul 2019   Prob (F-statistic):               0.00
Time:                        11:19:57   Log-Likelihood:                -7732.1
No. Observations:                4348   AIC:                         1.550e+04
Df Residuals:                    4328   BIC:                         1.563e+04
Df Model:                          19                                         
Covariance Type:                  HC1                                         
                             coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------------------
Intercept                 -1

In [246]:
sms.het_breuschpagan(fitted.resid, fitted.model.exog)[1]

1.1452927633437192e-225

Question 7: Ошибка гетероскедастична, p ≤ 0.05, нужно делать поправку Уайта

Удалите из модели незначимые признаки religion, radio и tv. Проверьте гомоскедастичность ошибки, при необходимости сделайте поправку Уайта.

Не произошло ли значимого ухудшения модели после удаления этой группы признаков? Проверьте с помощью критерия Фишера. Чему равен его достигаемый уровень значимости? Округлите до четырёх цифр после десятичной точки.

Если достигаемый уровень значимости получился маленький, верните все удалённые признаки; если он достаточно велик, оставьте модель без религии, тв и радио.

In [247]:
m3 = smf.ols('ceb ~ age + educ + idlnchld + knowmeth + usemeth + agefm + heduc + urban + electric +'\
                    'bicycle + nevermarr + idlnchld_noans + heduc_noans + usemeth_noans', 
             data=data)
fitted = m3.fit(cov_type='HC1')

In [248]:
sms.het_breuschpagan(fitted.resid, fitted.model.exog)[1]

1.1197458896531511e-228

In [249]:
m2.fit().compare_f_test(m3.fit())

(0.9192357784633435, 0.46723055472730823, 5.0)

In [250]:
round(m2.fit().compare_f_test(m3.fit())[1], 4)

0.4672

Question 8: 0.4672

Признак usemeth_noans значим по критерию Стьюдента, то есть, при его удалении модель значимо ухудшится. Но вообще-то отдельно его удалять нельзя: из-за того, что мы перекодировали пропуски в usemeth произвольно выбранным значением −1, удалять usemeth_noans и usemeth можно только вместе.

Удалите из текущей модели(Текущая это последняя или лучшая? m2 или m3? m2 - лучшая исходя из предыдущего вопроса) usemeth_noans и usemeth. Проверьте критерием Фишера гипотезу о том, что качество модели не ухудшилось. Введите номер первой значащей цифры в достигаемом уровне значимости (например, если вы получили 5.5×10^−8, нужно ввести 8).

Если достигаемый уровень значимости получился маленький, верните удалённые признаки; если он достаточно велик, оставьте модель без usemeth и usemeth_noans. 

In [257]:
m4 = smf.ols('ceb ~ age + educ + idlnchld + knowmeth + agefm + heduc + urban + electric +'\
                    'bicycle + nevermarr + idlnchld_noans + heduc_noans', 
             data=data)
fitted = m4.fit(cov_type='HC1')
print fitted.summary()

                            OLS Regression Results                            
Dep. Variable:                    ceb   R-squared:                       0.629
Model:                            OLS   Adj. R-squared:                  0.628
Method:                 Least Squares   F-statistic:                     396.4
Date:                Wed, 03 Jul 2019   Prob (F-statistic):               0.00
Time:                        11:27:44   Log-Likelihood:                -7825.7
No. Observations:                4348   AIC:                         1.568e+04
Df Residuals:                    4335   BIC:                         1.576e+04
Df Model:                          12                                         
Covariance Type:                  HC1                                         
                     coef    std err          z      P>|z|      [0.025      0.975]
----------------------------------------------------------------------------------
Intercept         -1.1931      0.262     -4.

In [258]:
m3.fit().compare_f_test(m4.fit())

(92.89058230109757, 3.1552009480386492e-40, 2.0)

Question 9: 40

In [273]:
fitted = m3.fit(cov_type='HC1')
print fitted.summary()

                            OLS Regression Results                            
Dep. Variable:                    ceb   R-squared:                       0.644
Model:                            OLS   Adj. R-squared:                  0.643
Method:                 Least Squares   F-statistic:                     463.4
Date:                Wed, 03 Jul 2019   Prob (F-statistic):               0.00
Time:                        11:44:29   Log-Likelihood:                -7734.5
No. Observations:                4348   AIC:                         1.550e+04
Df Residuals:                    4333   BIC:                         1.559e+04
Df Model:                          14                                         
Covariance Type:                  HC1                                         
                     coef    std err          z      P>|z|      [0.025      0.975]
----------------------------------------------------------------------------------
Intercept         -1.0698      0.258     -4.

In [274]:
0.6565 + -1*0.0770

0.5795

Question 10: У женщин, никогда не выходивших замуж, при прочих равных в среднем на 2.3 ребёнка меньше (p<0.001, 95% доверительный интервал для разницы между средними — [-2.6, -1.9])
У женщин, не знающих, какое количество детей идеально, в среднем на 0.66 ребёнка больше (p=0.002, 95% доверительный интервал — [0.2, 1.1])
С увеличением возраста женщины на 1 год среднее количество детей возрастает на 0.17 (p<0.001, 95% доверительный интервал — [0.16, 0.18])
У женщин, не знающих, какое количество детей идеально, в среднем на
βidlnchld_noans+cidlnchld*βidlnchld≈0.58
детей больше